core/lib/multivm/src/versions/vm_latest/tracers/circuits_tracer.rs

```diff
@@ -0,0 +1,196 @@
+            .history();
+        for (_, history_event) in &history[last_decommitment_history_entry_checked..] {
+            // We assume that only insertions may happen during a single VM inspection.
+            assert!(history_event.value.is_some());
+            let bytecode_len = state
+                .decommittment_processor
+                .known_bytecodes
+                .inner()
+                .get(&history_event.key)
+                .expect("Bytecode must be known at this point")
+                .len();
+
+            // Each cycle of `CodeDecommitter` processes 64 bytes.
+            let decommitter_cycles_used = bytecode_len / 4;
+            self.estimated_circuits_used +=
+                (decommitter_cycles_used as f32) * CODE_DECOMMITTER_CYCLE_FRACTION;
+        }
+        self.last_decommitment_history_entry_checked = Some(history.len());
+
+        // Process storage writes.
+        let last_writes_history_entry_checked = self
+            .last_written_keys_history_entry_checked
+            .expect("Value must be set during init");
+        let history = state.storage.written_keys.history();
+        for (_, history_event) in &history[last_writes_history_entry_checked..] {
+            // We assume that only insertions may happen during a single VM inspection.
+            assert!(history_event.value.is_some());
+
+            self.estimated_circuits_used += 2.0 * STORAGE_APPLICATION_CYCLE_FRACTION;
+        }
+        self.last_written_keys_history_entry_checked = Some(history.len());
+
+        // Process storage reads.
+        let last_reads_history_entry_checked = self
+            .last_read_keys_history_entry_checked
+            .expect("Value must be set during init");
+        let history = state.storage.read_keys.history();
+        for (_, history_event) in &history[last_reads_history_entry_checked..] {
+            // We assume that only insertions may happen during a single VM inspection.
+            assert!(history_event.value.is_some());
+
+            // If the slot is already written to, then we've already taken 2 cycles into account.
+            if !state
+                .storage
+                .written_keys
+                .inner()
+                .contains_key(&history_event.key)
+            {
+                self.estimated_circuits_used += STORAGE_APPLICATION_CYCLE_FRACTION;
+            }
+        }
+        self.last_read_keys_history_entry_checked = Some(history.len());
+
+        // Process precompiles.
+        let last_precompile_history_entry_checked = self
+            .last_precompile_history_entry_checked
+            .expect("Value must be set during init");
+        let history = state
+            .precompiles_processor
+            .precompile_cycles_history
+            .history();
+        for (_, history_event) in &history[last_precompile_history_entry_checked..] {
+            if let VectorHistoryEvent::Push((precompile, cycles)) = history_event {
```

Comment:
nit: in the `for` cycles above `assert` was used to assure the correctness of the execution, maybe also use it here